Ensure design docs are uploaded individually when replicating with bulk_get

nickva opened the pull request:
Previously, when replication jobs used _bulk_get they didn't upload design docs individually like they do when not using _bulk_get.

Here were are preserving an already existing behavior which we had in the replicator without _bulk_get usage for more than 2 years. It was introduced here in https://github.com/apache/couchdb/pull/2426. Related to these issues https://github.com/apache/couchdb/issues/2415 and https://github.com/apache/couchdb/issues/2413.

Add tests to cover both attachments and ddoc cases. meck:num_calls/3 is helpful as it allows to nicely assert which API function was called and how many times.

rnewson commented:
why?

rnewson reviewed (changes requested):
why?

There's no reason I can see to artificially limiting how many design documents are posted in one _bulk_get. Other clients are under no obligation to do the same, so any assumption you might make elsewhere, that ddocs are updated individually is built on a shaky foundation.

Please explain why this is an improvement, or bug fix, for couchdb on its own terms.

jaydoane reviewed (approved):
This seems like an elegant way to prevent using replication to exceed ddoc limits.

nickva commented:
@rnewson good question, I should have explain that a bit better in the commit.

That is to preserve an already existing behavior which we had in the replicator without _bulk_get usage for more than 2 years. It was introduced here: https://github.com/apache/couchdb/pull/2426 related to these issues https://github.com/apache/couchdb/issues/2415 and https://github.com/apache/couchdb/issues/2413

nickva commented on src/couch_replicator/src/couch_replicator_worker.erl:
```diff
@@ -349,8 +349,9 @@ bulk_get(#httpdb{} = Source, #{} = IdRevs) ->
         {ok, #{} = Docs} ->
             FilterFun = fun
-                (_, #doc{atts = []}) -> true;
+                (_, #doc{id = <<?DESIGN_DOC_PREFIX, _/binary>>}) -> false;
```
I see minor optimization here - since we know right off the bat these are design documents, we don't have to fetch them in bulk_get at all (unlike the ones with attachments). So am thinking of updating to fetch and upload design docs, fetch the bulk docs, do the attachments ones. What do we think?

nickva replied:
Updated as described in the note

nickva commented:
Updated with the minor optimization and added references in the commit message and PR comment to the previous PR which introduced this behavior and the related issue.

rnewson commented:
@nickva thanks for the links. It still seems to be a (perhaps necessary) hack around a problem elsewhere. We should do The Right Thing for any _bulk_doc write that writes multiple design documents. It is not right imo to just teach our replicator to avoid triggering those bugs, no other client will be similarly protected.

nickva commented:
@rnewson with the replicator we are trying to be compatible even with old, or possibly broken clients, clients. But most of all, in this PR, we are trying to bring back the previous behavior before the recent _bulk_get optimization. Currently, it's odd that the there is a difference based on whether `[replicator] use_bulk_get` is toggled or not.
